### PR TITLE
Manual Eigen lstm module

### PR DESCRIPTION
**Pull request by iliaeg**
- Implemented lstm eigen module

### CODE REVIEW

**Comment by iliaeg** [comment]
LSTM  Jacobian  - Release Graph
LSTM  Jacobian ÷ objective  - Release Graph
LSTM  Objective  - Release Graph
LSTM  Objective ÷ manual  - Release Graph

**Comment by iliaeg** [comment]
- [ ] Tests on linux passed

File: src/cpp/modules/manualEigen/lstm_eigen_d_structures.h [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+#include <vector>
+
+#include "../../shared/lstm_eigen_types.h"
+
+template<typename T>
+struct StateElementGradientModel
+{
+    // Matrix(hsize, 10) derivatives from all layers
+    MapX10<T> d_rawX10;
```
Comment (msdmkats) [inline src/cpp/modules/manualEigen/lstm_eigen_d_structures.h]: Here and below `MapXN` varaiables need explanations in comments: what exactly are they (which derivatives they incorporate) and why do we need them.

File: src/cpp/modules/manualEigen/lstm_eigen_d.cpp [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    state.cell = orig_cell * forget + ingate * change;
+    // wrt weight
+    jacobian.cell.d_weight_forget = orig_cell * forget_dw;
+    jacobian.cell.d_weight_ingate = change * ingate_dw;
+    jacobian.cell.d_weight_outgate = 0.;
+    jacobian.cell.d_weight_change = ingate * change_dw;
+    // wrt bias
+    jacobian.cell.d_bias_forget = orig_cell * forget_db;
+    jacobian.cell.d_bias_ingate = change * ingate_db;
+    jacobian.cell.d_bias_outgate.setZero();
```
Comment (msdmkats) [inline src/cpp/modules/manualEigen/lstm_eigen_d.cpp]: Why do we use use `= 0.` for `d_weight_outgate` and `.setZero()` for `d_bias_outgate`? Is there any difference?

Comment (iliaeg) [reply src/cpp/modules/manualEigen/lstm_eigen_d.cpp]: No, its not. I've just overlooked that.
Fixed now.

File: src/cpp/modules/manualEigen/lstm_eigen_d_structures.h (hunk 1 above)
Comment (iliaeg) [reply src/cpp/modules/manualEigen/lstm_eigen_d_structures.h]: Comments added.

**Pull request merged**